firebase-sessions/src/main/kotlin/com/google/firebase/sessions/ProcessUtils.kt

```diff
@@ -0,0 +1,84 @@
+ * limitations under the License.
+ */
+
+package com.google.firebase.sessions
+
+import android.app.ActivityManager
+import android.app.ActivityManager.RunningAppProcessInfo
+import android.app.Application
+import android.content.Context
+import android.os.Build
+import android.os.Process
+
+/** Utils related to processes. */
+internal object ProcessUtils {
+  /** Returns whether the current process should generate a new session or not. */
+  fun shouldProcessGenerateNewSession(context: Context): Boolean =
+    isDefaultProcess(context) && isForegroundImportance()
+
+  /** Returns whether the current process is the app's default process or not. */
+  private fun isDefaultProcess(context: Context): Boolean =
+    getCurrentProcessName(context) == getDefaultProcessName(context)
+
+  /** Returns whether the current process or service is running in the foreground or not. */
+  private fun isForegroundImportance(): Boolean {
+    val runningAppProcessInfo = RunningAppProcessInfo()
+    ActivityManager.getMyMemoryState(runningAppProcessInfo)
+    return isForegroundProcess(runningAppProcessInfo) || isForegroundService(runningAppProcessInfo)
+  }
+
+  /** Returns the name of the current process. */
+  private fun getCurrentProcessName(context: Context): String? =
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
+      Application.getProcessName()
+    } else {
+      findProcessName(context, Process.myPid())
+    }
+
+  /** Finds the process name for the given pid, or returns null if not found. */
+  private fun findProcessName(context: Context, pid: Int): String? {
+    val activityManager = context.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
+    return activityManager.runningAppProcesses?.find { it.pid == pid }?.processName
+  }
+
+  /**
+   * Gets the default process name.
+   *
+   * This is the app's package name unless the app overrides the android:process attribute in the
+   * application block of its Android manifest file.
+   */
+  private fun getDefaultProcessName(context: Context): String = context.applicationInfo.processName
+
+  /**
+   * Returns if the current process is at the top of the screen that the user is interacting with.
+   */
+  private fun isForegroundProcess(runningAppProcessInfo: RunningAppProcessInfo): Boolean {
+    return runningAppProcessInfo.importance == RunningAppProcessInfo.IMPORTANCE_FOREGROUND
+  }
+
+  /**
+   * Returns if the current process is running as a foreground service.
+   *
+   * This generally indicates that the process is doing something the user actively cares about.
+   */
+  private fun isForegroundService(runningAppProcessInfo: RunningAppProcessInfo): Boolean =
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+      runningAppProcessInfo.importance == RunningAppProcessInfo.IMPORTANCE_FOREGROUND_SERVICE
+    } else {
+      // Default to false if the api level is too low to check.
+      false
```

Comment:
What happens in these scenarios? We would not generate a session?

Reply:
Correct. If  shouldProcessGenerateNewSession == false, then we do not generate a session. :)

Reply:
Interesting - in that case, any lower API versions, would we be not able to generate sessions? Should we record this as a limitation? Do we know the number of instances in this API version? We don't have to go deeper to this. But getting a sense of the limitation would be helpful to make a decision.

Reply:
Unless I'm missing something, the code checks if it's a foreground process OR if it's a foreground service. So, if it's unable to determine if it's a foreground service, it will still generate a session if it's a foreground process. Is this your take as well?